File: src/main/java/org/lova/model/RiceCooker.java
package org.lova.model;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.UUID;

public class RiceCooker {
  private String id;
  private Status status;
  private Cook cook;

  public RiceCooker() {
    this.id = UUID.randomUUID().toString();
    this.status = Status.OFF;
    this.cook = null;
  }

  public void startCooking(String recipeName, String recipeDescription, Integer duration) {
    cook = new Cook(recipeName, recipeDescription, duration);
    status = Status.COOKING;
  }

  public void stopCooking() {
    cook = null;
    status = Status.OFF;
  }

  public void updateStatus(Instant currentDatetime) {
    if (cook != null) {
      Instant endDateTime = getEndDatetime(cook.getDebut(), cook.getDuration());

      if (endDateTime.isBefore(currentDatetime)) {
        status = Status.WARMING;
      }
    }
  }

  public void showStatus() {
    Instant now = Instant.now();
    updateStatus(now);
    System.out.println();
    System.out.println("Status: " + status);
    if (cook == null) {
      System.out.println("No cooking has been done");
    } else {
      if (status == Status.COOKING) {
        System.out.println("Recipe: " + cook.getRecipeName());
        System.out.println("Time remains: " + -getEndDatetime(cook.getDebut(), cook.getDuration()).until(now,
            ChronoUnit.MINUTES) + " MIN");
      } else if (status == Status.WARMING) {
        System.out.println("Cooking done!");
      }
    }
  }

  private Instant getEndDatetime(Instant debut, Integer duration) {
    long durationInSecond = duration * 60;
    return debut.plusSeconds(durationInSecond);
  }

  public static enum Status {
    OFF, COOKING, WARMING
  }

  public String getId() {
    return id;
  }

  public Status getStatus() {
    return status;
  }

  public Cook getCook() {
    return cook;
  }
}

File: RiceCooker.py
from datetime import datetime, timedelta
from Cook import Cook
import uuid


class RiceCooker:
    class Status:
        OFF = "OFF"
        COOKING = "COOKING"
        WARMING = "WARMING"

    def __init__(self):
        self.id = str(uuid.uuid4())
        self.status = self.Status.OFF
        self.cook = None

    def start_cooking(self, recipe_name, recipe_description, duration):
        self.cook = Cook(recipe_name, recipe_description, duration)
        self.status = self.Status.COOKING

    def stop_cooking(self):
        self.cook = None
        self.status = self.Status.OFF

    def update_status(self):
        if self.cook is not None:
            end_datetime = self.cook.debut + timedelta(minutes=self.cook.duration)

            if end_datetime < datetime.now():
                self.status = self.Status.WARMING

    def show_status(self):
        self.update_status()
        now = datetime.now()
        print()
        print("Status:", self.status)

        if self.cook is None:
            print("No cooking has been done")
        else:
            if self.status == self.Status.COOKING:
                time_remaining = (
                    self.cook.debut + timedelta(minutes=self.cook.duration) - now
                ).seconds // 60
                print("Recipe:", self.cook.recipe_name)
                print("Time remains:", time_remaining, "MIN")
            elif self.status == self.Status.WARMING:
                print("Cooking done!")
